src/main/java/com/epam/controller/TraineeController.java

```diff
@@ -0,0 +1,153 @@
+
+    @PostMapping
+    public ResponseEntity<TraineeRegistrationResponse> registerTrainee(@RequestBody @Valid TraineeRegistrationRequest request) {
+        String username = UserUtils.getBaseUsername(request.getFirstname(), request.getLastname());
+        Optional<Trainer> trainerOptional = trainerService.getTrainerByUsername(username);
+        if (trainerOptional.isPresent()) {
+            ResponseEntity.badRequest().body("Not possible to register as a trainer and trainee both");
+        }
+
+        Trainee trainee = mapper.traineeRegistrationRequestToTrainee(request);
+        Trainee createdTrainee = traineeService.createTrainee(trainee);
+        TraineeRegistrationResponse response = new TraineeRegistrationResponse(
+                createdTrainee.getUser().getUsername(),
+                createdTrainee.getUser().getPassword());
+        return ResponseEntity.ok(response);
+    }
+
+    @GetMapping
+    @Transactional
+    public ResponseEntity<TraineeProfileResponse> getTraineeProfile(@RequestParam String username) {
+        Optional<Trainee> traineeOptional = traineeService.getTraineeByUsername(username);
+        if (traineeOptional.isEmpty()) {
+            ResponseEntity.notFound();
+        }
+        Trainee trainee = traineeOptional.get();
+        TraineeProfileResponse profile = mapper.traineeToTraineeProfileResponse(trainee);
+
+        return ResponseEntity.ok(profile);
+    }
+
+    @PutMapping
+    @Transactional
+    public ResponseEntity<TraineeUpdateResponseDto> updateTraineeProfile(@RequestParam(name = "username") String username,
+                                                                         @Valid @RequestBody TraineeUpdateRequestDto request) {
+        Optional<Trainee> updatedTraineeOptional = traineeService.getTraineeByUsername(username);
+        if (updatedTraineeOptional.isEmpty()) {
+            return ResponseEntity.notFound().build();
+        }
+        Trainee updatedTrainee = updatedTraineeOptional.get();
+        mapper.updateEntity(updatedTrainee, request);
+
+        Trainee trainee = traineeService.updateTrainee(updatedTrainee).get();
+        TraineeUpdateResponseDto updateResponseDto = mapper.traineeToTraineeUpdateResponseDto(trainee);
+        return ResponseEntity.ok(updateResponseDto);
+    }
+
+    @DeleteMapping
+    public ResponseEntity<String> deleteTraineeProfile(@RequestParam String username) {
+        try {
+            boolean deleted = traineeService.deleteTraineeByUsername(username);
+
+            if (deleted) {
+                return ResponseEntity.ok("Trainee profile deleted successfully");
+            } else {
+                return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Trainee not found");
+            }
+        } catch (DataAccessException ex) {
+            log.error("Error deleting trainee profile", ex);
+            throw new OperationFailureException("Error deleting trainee profile. Please try again later.");
```

Comment:
Here it seems inappropriate to throw the Exception because we can just return the corresponding bad httpStatus, right? Instead, you should have moved this try-catch within the `deleteTraineeByUsername` method of the service layer. And the exception thrown from the service layer will be caught by your high-level handler.